### PR TITLE
[Enhancement] improve performance of array_agg function

### PR DESCRIPTION
**Pull request by murphyatwork**
## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

Improve the performance of `array_agg` function:
1. Eliminate virtual function call
2. Eliminate the `datum` intermediate representation

Benchmark Result:
1. `select * from (select lo_orderkey, array_agg(lo_orderkey) as agg from lineorder group by lo_orderkey) r order by lo_orderkey limit 1` 
    1. Before: `3s369ms`
    2. After: `2s958ms`
2. `select * from (select lo_custkey, array_agg(lo_custkey) as agg from lineorder group by lo_custkey) r order by lo_custkey limit 1`
    1. Before: `5s596ms `
    2. After: `3s760ms`
3. The first sql generate array whose max length is 7, and the second is 611. For large array, this PR has significant improment

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

### CODE REVIEW

**Review by ZiheLiu — DISMISSED** [review]

File: be/src/exprs/agg/array_agg.h [hunk 1]
```diff
@@ -3,88 +3,61 @@
...
-        auto& elements_column = column->elements_column();
-
-        if (null_count > 0) {
-            elements_column->append_nulls(null_count);
-        }
-        for (size_t i = 0; i < data.size(); i++) {
-            elements_column->append_datum(data[i]);
-        }
-        offsets.emplace_back(offsets.back() + data.size() + null_count);
+        column->append_array_element(state_impl.data_column, state_impl.null_count);
```
Comment (LiShuMing) [inline be/src/exprs/agg/array_agg.h:60]: Does `deserialize` is supported for  ArrayAGG ?

**Review by satanson — DISMISSED** [review]

**Comment by ZiheLiu** [comment]
run starrocks_admit_test

File: be/src/column/nullable_column.cpp [hunk 2]
```diff
@@ -113,7 +120,9 @@ ColumnPtr NullableColumn::replicate(const std::vector<uint32_t>& offsets) {
 }
 
 bool NullableColumn::append_nulls(size_t count) {
-    DCHECK_GT(count, 0u);
+    if (count == 0) {
```
Comment (kangkaisen) [inline be/src/column/nullable_column.cpp:123]: why check count in here?

File: be/src/exprs/agg/array_agg.h (hunk 1 above)
Comment (kangkaisen) [reply be/src/exprs/agg/array_agg.h:60]: deserialize is done in merge method.  deserialize and merge at the same time.

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by mofeiatwork** [comment]
@Mergifyio rebase

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

**Comment by mofeiatwork** [comment]
run starrocks_be_unittest

**Comment by mofeiatwork** [comment]
run starrocks_admit_test

File: be/src/column/nullable_column.cpp (hunk 2 above)
Comment (mofeiatwork) [reply be/src/column/nullable_column.cpp:123]: Otherwise every caller need to check it.

**Comment by fzhedu** [comment]
are there some tests for it?

**Review by stdpain — APPROVED** [review]

File: be/src/column/nullable_column.cpp (hunk 2 above)
Comment (kangkaisen) [reply be/src/column/nullable_column.cpp:123]: I think caller check is also OK. This change will effect the exist caller performance.

**Review by ZiheLiu — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

**Pull request merged**